README
Delay - README file
===================

Tom Rothamel <tom@onegeek.org>

This program implements a way to implement a delay with feedback to the 
user. It supports a configurable output message, and tries to keep the 
timing precise. 

Please read the INSTALL file for information on how to compile and 
install this program. Once installed, 'man delay' should display 
information on how to use the program.

The parsetimel.l and parsetimey.y files were taken from the files 
parsetime.l and parsetime.y (respectively) from the at-3.1.7 distribution,
which was authored by Thomas Koenig and David Parsons. These files are
used to implement the "delay until" feature. 

Thanks are due to:

 - Gianmarco Giovannelli <gmarco@giovannelli.it>
	For maintaining a FreeBsd port of delay, and working with me on
	resolving a few compatibility issues.
 - Tomi Ollila <Tomi.Ollila@tfi.net>
	For suggesting many of the features added to version 1.2, and
	helping with Solaris compatiblity.
 - Zachary P. Landau <kapheine@www.annex.org> 
	For suggesting giving delay the ability to run a command, 
	which was added in version 1.3.

- Tom Rothamel
  


ChangeLog
Version 1.6 - 09 February 2002
	Fixes a bug in the spelling of February in the time parsing 
	code. 

	Added support for c-style string escapes to the custom format
	parsing code. Also, removed the implicit carriage return before
	the printing of custom codes. This allows delay to integrate
	with programs like osd_cat, using a command line like:

	./delay -c '%02h:%02m:%02s\n' 1h | osd_cat

Version 1.5 - 19 August 2001 
	This version fixes a bug in getopt., and removes support for
	building debian packages and rpms. (Those are best done by the
	various distro maintainers, anyway.)

Version 1.4 - 30 December 1999
	This version adds support for big ugly curses numbers. This
	support can be enabled using the '-C' option.

Version 1.3 - 29 June 1999
	This version allows delay to take a command on the command 
	line which specifies a program to run when the countdown
	is complete. (This way, people no longer have to use a 
	construct like "delay 1m; command" which runs the command
	if delay is interrupted.)
	
Version 1.2a - 2 March 1999
	This version fixes support for Solaris, which was broken in
	the BSD fixes. 

Version 1.2 - 1 March 1999
	This version adds sleep-compatability and fractional-second 
	delaying, and is the first public release with the FreeBsd
	patches.

Version 1.1b - 24 February 1999
	This includes a minor rewrite in the gmt code, so that it works
	with FreeBsd. And this time it actually works!

Version 1.1a - 23 February 1999
	This includes a minor rewrite in the gmt code, so that it works
	with FreeBsd.

Version 1.1 - 31 January 1999
	This release adds support for a "delay until" feature, the ability
	to ring a bell when time runs out, and a changed default format

Version 1.0 - 26 January 1999
	Initial release.



curses.c
#ifdef HAVE_LIBCURSES
#ifdef HAVE_CURSES_H

#include <curses.h>

#define FWIDTH 6
#define FHEIGHT 7

char *digits[10][7] = {
	{
		" **** ",
		"**  **",
		"**  **",
		"**  **",
		"**  **",
		"**  **",
		" **** ",
	 }, {
		"   ** ",
		"  *** ",
		"   ** ",
		"   ** ",
		"   ** ",
		"   ** ",
		"   ** ",
	 }, {
		" **** ",
		"**  **",
		"    **",
		"   ** ",
		"  **  ",
		" **   ",
		"******",
	 }, {
		" **** ",
		"**  **",
		"    **",
		" **** ",
		"    **",
		"**  **",
		" **** ",
	 }, {
		"   ** ",
		"  *** ",
		" * ** ",
		"*  ** ",
		"******",
		"   ** ",
		"   ** ",
	 }, {
		"******",
		"**    ",
		"**    ",
		"***** ",
		"    **",
		"    **",
		"***** ",
	 }, {
		" **** ",
		"**  **",
		"**    ",
		"***** ",
		"**  **",
		"**  **",
		" **** ",
	 }, {
		"******",
		"    **",
		"   ** ",
		"  **  ",
		"  **  ",
		"  **  ",
		"  **  ",
	 }, {
		" **** ",
		"**  **",
		"**  **",
		" **** ",
		"**  **",
		"**  **",
		" **** ",
	 }, {
		" **** ",
		"**  **",
		"**  **",
		" *****",
		"    **",
		"**  **",
		" **** ",
	 }
};

void printrow(char *s) {
	while (*s) {
		switch(*s) {
		case '*':
			addch(' ' | A_REVERSE);
			break;
		default:
			addch(*s);
			break;
		}
		s++;
	}
}

void printdigit(int r, int c, int d) {
	int i;
	for(i=0; i < 7; i++) {
		move(r+i, c);
		printrow(digits[d][i]);
	}
}

void printcolon(int r, int c) {
	move(r+2, c);
	addch(' ' | A_REVERSE);
	move(r+5, c);
	addch(' ' | A_REVERSE);
}		

void printstring(char *s) {
	int col = 2;
	int row = 2;

	while (*s) {
		switch (*s) {
		case '0':
		case '1':
		case '2':
		case '3':
		case '4':
		case '5':
		case '6':
		case '7':
		case '8':
		case '9':
			printdigit(row, col, *s - '0');
			col += FWIDTH + 1;
			break;
		case ' ':
			col += FWIDTH + 1;
			break;
		case ':':
			printcolon(row, col);
			col += 2;
		}
		s++;
	}
}

void curshowcount(int n) {
	char buf[25];

	sprintf(buf, "%d:%02d:%02d", n / 3600,
		(n/60) % 60, n % 60);

	erase();
	printstring(buf);
	move(LINES-1, COLS-1);
	refresh();
}
	

#endif
#endif


parsetime.h
/* 
 *  at.h -  header for at(1)
 *  Copyright (C) 1993  Thomas Koenig
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

time_t parsetime(int argc, char **argv);

extern char *last_token;
extern char **my_argv;

#define NEED_YYWRAP



delay.c
/* delay - counts down a specified number of seconds.
   Copyright (C) 1998-2014 Tom Rothamel.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#include <stdio.h>
#include <time.h>
#include <sys/time.h>
#include <unistd.h>
#include <string.h>
#include <stdlib.h>
#include "parsetime.h"

#ifdef HAVE_CURSES_H
#include <curses.h>
#endif

/* The following code implements the actual delaying. If precise is true, it
   attempts to compensate for the time spend in calculating, printing, etc.
*/

struct timeval tv;

void initdelay(int precise) {
	if (!precise) return;
	gettimeofday(&tv, NULL);
}

void delay(int secs, int precise) {
	struct timeval ctv;
	struct timeval dtv;

	if (!precise) {
		sleep(secs);
		return;
	}

	tv.tv_sec += secs;

	gettimeofday(&ctv, NULL);
	dtv.tv_sec  = tv.tv_sec - ctv.tv_sec;
	dtv.tv_usec  = tv.tv_usec - ctv.tv_usec;

	if (dtv.tv_usec < 0) {
		dtv.tv_usec += 1000000;
		dtv.tv_sec -= 1;
	}

	select(0, NULL, NULL, NULL, &dtv);
}

/* These two functions  handle the parsing of standard-format arguments.
   This format allows for HH:MM:SS, DDdHHhMMmSSs, or SSSSs style inputs.
   (Multiple arguments are added together.
*/

int parseonestd(char *entry) {
	int commit = 0;
        int t = 0;
	int usfact = 1000000;

	while (1) {
		if (!*entry) return commit + t;


		if (('0' <= *entry) && (*entry <= '9')) {
			t *= 10;
			t += *entry - '0';
		} else if (*entry == '.') {
			break;
		} else switch(*entry) {
		case ':':
			commit +=t;
			commit *= 60;
			t = 0;
			break;
		case 's':
			commit += t;
			t = 0;
			break;
		case 'm':
			commit +=  t * 60;
			t = 0;
			break;
		case 'h':
			commit += t * 3600;
			t = 0;
			break;
		case 'd':
			commit += t * 86400;
			break;
		}

		entry++;
	}

	while (*++entry) {
		usfact /= 10;
		if (('0' <= *entry) && (*entry <= '9')) {
			usleep((*entry - '0') * usfact);
		}
	}

	return commit + t;
}


int parsestandard(int argc, char **argv) {
	int i;
	int dt = 0;
	int tdt;

	for (i = 0; i < argc; i++) {
		tdt = parseonestd(argv[i]);
		if (tdt == -1) return -1;
		dt += tdt;
	}

	return dt;
}



/* Prints out the time, using the format code. The format is somewhat like
   the integer codes used in printf. You have all the flags, and then
   d for days, h for hours, m for minutes, s for seconds, and n for the
   total number of seconds remaining.
*/

void printtime(char *fmt, int time) {
	char buf[16];
	char *c;
	int i;
	char b;

	c = fmt;


	while (*c) {
		switch (*c) {
		case '\\':
			c++;

			switch (*c) {
			case 0:
				return;
			case '\\':
				printf("\\");
				break;
			case 'r':
				printf("\r");
				break;
			case 'n':
				printf("\n");
				break;
			case 't':
				printf("\t");
				break;
			case '%':
				printf("%%");
				break;
			}

			break;

		case '%':
			strncpy(buf, c, 15);
			for (i = 1; i < 100; i++) {
				 if (buf[i] == 0) return;
				 if (buf[i] == 'd') break;
				 if (buf[i] == 'h') break;
				 if (buf[i] == 'm') break;
				 if (buf[i] == 's') break;
				 if (buf[i] == 'n') break;
			}

			b = buf[i];
			buf[i] = 'd';
			buf[i+1] = 0;

			switch (b) {
			case 'd':
				printf(buf, time/86400);
				break;
			case 'h':
				printf(buf, (time/3600) % 24);
				break;
			case 'm':
				printf(buf, (time/60) % 60);
				break;
			case 's':
				printf(buf, time % 60);
				break;
			case 'n':
				printf(buf, time);
				break;
			case 0:
				return;
			}

			c += i;
			break;
		default:
			printf("%c", *c);
			break;
		}

		c++;
	}

}


/* The rest of the code is responsible for displaying the countdown,
   parsing the arguments, and otherwise making this thing work.
*/

char *custmessage;

void usage(char *name) {
	fprintf(stdout,
		"Usage: %s [options] <seconds to delay> [-- command]\n"
		"   or  %s [options] until <timespec> [-- command]\n", name, name);

	exit(1);
}

void curshowcount(int);

void showcount(int dtime, int ctype) {
	switch(ctype) {
	case 0:
		printtime("\r% 3d %02h:%02m:%02s", dtime);
		break;
	case 1:
		printtime("\rTime Remaining: %d days, %02h:%02m:%02s.", dtime);
		break;
	case 2:
		printtime(custmessage, dtime);
		break;
	case 3:
		printf("\r% 8d", dtime);
		break;
	case 4:
		curshowcount(dtime);
		break;

	}

	fflush(stdout);
}

int main(int argc, char **argv) {
	int opt;
	int dtime;
	int ctype = 0;
	int bell = 0;
	int update = 1;
	int i;
	char **cmd = NULL;


	if (strstr(argv[0], "sleep")) ctype = -1;

	for (i = 0; i < argc; i++) {
		if(!strcmp(argv[i], "--")) {
			if (i == argc - 1) {
				fprintf(stderr, "%s: with '--' you must specify a command to run.\n", argv[0]);
				usage(argv[0]);
			}

			cmd = &argv[i+1];
			argv[i] = NULL;
			argc = i;
			break;
		}
	}

	while (1) {
		opt = getopt(argc, argv, "qu:bhdvmc:CV");
		if (opt == EOF) break;

		switch (opt) {
		case 'q':
			ctype = -1;
			break;
		case 'u':
			update = atoi(optarg);
			break;
		case 'h':
			usage(argv[0]);
			break;
		case 'd':
			ctype = 0;
			break;
		case 'v':
		        ctype = 1;
			break;
		case 'm':
			ctype = 3;
			break;
		case 'c':
			custmessage = optarg;
			ctype=2;
			break;
		case 'C':
			ctype = 4;
			break;
		case 'V':
			printf("delay - Version " VERSION "\n"
			       "  Copyright (c) 1998-2002 Tom Rothamel\n"
			       "  This program has ABSOLUTELY NO WARRANTY. It can be distributed under\n"
			       "  the terms of the GNU General Public License.\n");
			exit(0);
		case 'b':
			bell = 1;
			break;
		}

	}

	if (optind >= argc) {
		fprintf(stderr, "%s: You must supply a time to delay.\n", argv[0]);
		usage(argv[0]);
	}

	if (update < 1) {
		fprintf(stderr, "%s: The update rate must be a positive integer.\n", argv[0]);
		exit(1);
	}

	if (!strcmp(argv[optind], "until")) {
		dtime = parsetime(argc-optind, &argv[optind+1]);
		if (!dtime) exit(1);
		dtime -= time(NULL);
	} else {
		dtime = parsestandard(argc-optind, &argv[optind]);
	}

	if (dtime < 0) {
		fprintf(stderr, "%s: You've specified an invalid delay time.\n", argv[0]);
		usage(argv[0]);
	}

	initdelay(1);

	if (ctype == 4) initscr();

	while (dtime > 0) {
		if (ctype >= 0) showcount(dtime, ctype);

		delay((dtime < update) ? dtime : update, 1); /* always precise. */
		dtime -= (dtime < update) ? dtime : update;
	}

	if (ctype >= 0) {
		showcount(dtime, ctype);
		if (ctype != 4) printf("\n");
	}

	if (ctype == 4) endwin();

	if (bell) printf("\a");

	if (cmd) {
		execvp(cmd[0], cmd);
		perror(argv[0]); /* If it worked, we won't get here. */
		exit(1);
	}

	exit(0);
}
